File: nodejs/src/validators/cpf.ts
export class CPFValidator {
  /**
   * Validate CPF format and check digits
   */
  isValid(cpf: string): boolean {
    if (!cpf || typeof cpf !== 'string') return false;
    
    const cleanCPF = this.removeMask(cpf);
    
    // Check if it has 11 digits
    if (cleanCPF.length !== 11) return false;
    
    // Check if all digits are the same
    if (/^(\d)\1{10}$/.test(cleanCPF)) return false;
    
    // Validate first check digit
    let sum = 0;
    for (let i = 0; i < 9; i++) {
      sum += parseInt(cleanCPF[i]) * (10 - i);
    }
    let remainder = sum % 11;
    let firstDigit = remainder < 2 ? 0 : 11 - remainder;
    
    if (parseInt(cleanCPF[9]) !== firstDigit) return false;
    
    // Validate second check digit
    sum = 0;
    for (let i = 0; i < 10; i++) {
      sum += parseInt(cleanCPF[i]) * (11 - i);
    }
    remainder = sum % 11;
    let secondDigit = remainder < 2 ? 0 : 11 - remainder;
    
    return parseInt(cleanCPF[10]) === secondDigit;
  }

  /**
   * Generate a valid CPF
   */
  generate(): string {
    const digits = Array.from({ length: 9 }, () => Math.floor(Math.random() * 10));
    
    // Calculate first check digit
    let sum = 0;
    for (let i = 0; i < 9; i++) {
      sum += digits[i] * (10 - i);
    }
    let remainder = sum % 11;
    let firstDigit = remainder < 2 ? 0 : 11 - remainder;
    digits.push(firstDigit);
    
    // Calculate second check digit
    sum = 0;
    for (let i = 0; i < 10; i++) {
      sum += digits[i] * (11 - i);
    }
    remainder = sum % 11;
    let secondDigit = remainder < 2 ? 0 : 11 - remainder;
    digits.push(secondDigit);
    
    return this.applyMask(digits.join(''));
  }

  /**
   * Apply CPF mask (000.000.000-00)
   */
  applyMask(cpf: string): string {
    if (!cpf || typeof cpf !== 'string') return '';
    
    const cleanCPF = this.removeMask(cpf);
    if (cleanCPF.length !== 11) return cpf;
    
    return cleanCPF.replace(/(\d{3})(\d{3})(\d{3})(\d{2})/, '$1.$2.$3-$4');
  }

  /**
   * Remove CPF mask
   */
  removeMask(cpf: string): string {
    if (!cpf || typeof cpf !== 'string') return '';
    return cpf.replace(/\D/g, '');
  }

  /**
   * Get state from CPF first two digits
   */
  getState(cpf: string): string | null {
    if (!this.isValid(cpf)) return null;
    
    const cleanCPF = this.removeMask(cpf);
    const firstTwoDigits = parseInt(cleanCPF.substring(0, 2));
    
    const stateMap: { [key: number]: string } = {
      1: 'DF, GO, MS, MT, TO',
      2: 'AC, AP, AM, PA, RO, RR',
      3: 'CE, MA, PI',
      4: 'AL, PB, PE, RN',
      5: 'BA, SE',
      6: 'MG',
      7: 'ES, RJ',
      8: 'SP',
      9: 'PR, SC',
      0: 'RS'
    };
    
    for (const [range, states] of Object.entries(stateMap)) {
      if (firstTwoDigits >= parseInt(range) * 10 && firstTwoDigits <= parseInt(range) * 10 + 9) {
        return states;
      }
    }
    
    return null;
  }
} 

File: python/validbr/validators/cpf.py
import random
import re
from typing import Optional


class CPFValidator:
    """CPF (Cadastro de Pessoas Físicas) validator."""
    
    def __init__(self):
        self.state_map = {
            1: 'DF, GO, MS, MT, TO',
            2: 'AC, AP, AM, PA, RO, RR',
            3: 'CE, MA, PI',
            4: 'AL, PB, PE, RN',
            5: 'BA, SE',
            6: 'MG',
            7: 'ES, RJ',
            8: 'SP',
            9: 'PR, SC',
            0: 'RS'
        }

    def is_valid(self, cpf: str) -> bool:
        """Validate CPF format and check digits."""
        if not cpf or not isinstance(cpf, str):
            return False
        
        clean_cpf = self.remove_mask(cpf)
        
        # Check if it has 11 digits
        if len(clean_cpf) != 11:
            return False
        
        # Check if all digits are the same
        if len(set(clean_cpf)) == 1:
            return False
        
        # Validate first check digit
        sum_val = sum(int(clean_cpf[i]) * (10 - i) for i in range(9))
        remainder = sum_val % 11
        first_digit = 0 if remainder < 2 else 11 - remainder
        
        if int(clean_cpf[9]) != first_digit:
            return False
        
        # Validate second check digit
        sum_val = sum(int(clean_cpf[i]) * (11 - i) for i in range(10))
        remainder = sum_val % 11
        second_digit = 0 if remainder < 2 else 11 - remainder
        
        return int(clean_cpf[10]) == second_digit

    def generate(self) -> str:
        """Generate a valid CPF."""
        digits = [random.randint(0, 9) for _ in range(9)]
        
        # Calculate first check digit
        sum_val = sum(digits[i] * (10 - i) for i in range(9))
        remainder = sum_val % 11
        first_digit = 0 if remainder < 2 else 11 - remainder
        digits.append(first_digit)
        
        # Calculate second check digit
        sum_val = sum(digits[i] * (11 - i) for i in range(10))
        remainder = sum_val % 11
        second_digit = 0 if remainder < 2 else 11 - remainder
        digits.append(second_digit)
        
        return self.apply_mask(''.join(map(str, digits)))

    def apply_mask(self, cpf: str) -> str:
        """Apply CPF mask (000.000.000-00)."""
        if not cpf or not isinstance(cpf, str):
            return ""
        
        clean_cpf = self.remove_mask(cpf)
        if len(clean_cpf) != 11:
            return cpf
        
        return f"{clean_cpf[:3]}.{clean_cpf[3:6]}.{clean_cpf[6:9]}-{clean_cpf[9:]}"

    def remove_mask(self, cpf: str) -> str:
        """Remove CPF mask."""
        if not cpf or not isinstance(cpf, str):
            return ""
        return re.sub(r'\D', '', cpf)

    def get_state(self, cpf: str) -> Optional[str]:
        """Get state from CPF first two digits."""
        if not self.is_valid(cpf):
            return None
        
        clean_cpf = self.remove_mask(cpf)
        first_two_digits = int(clean_cpf[:2])
        
        for range_start, states in self.state_map.items():
            if range_start * 10 <= first_two_digits <= range_start * 10 + 9:
                return states
        
        return None 